File: agent-driven-architecture-main/examples/kotlin/block/triage/src/main/kotlin/adr/blocks/triage/Fold.kt
// ── blocks/triage/fold — the block's ARM ───────────────────────────────────
// Exhaustive over the block's sealed sub-union, with NO else arm. Three rules (§7),
// mechanical, no exceptions:
//
//   1. the arm READS CURRENT STATE before it decides;
//   2. every effect push lives INSIDE THE SUCCESS BRANCH;
//   3. a rejection folds a per-item Notice.Rejected — never RunStatus, never a
//      mutation.
//
// the review's measured (12.4) bug lived exactly here: setPriority on unknown ticket 9999 performed
// an Effect.Log, committed a SetPriority command and left the folded state
// UNCHANGED — a clean-looking audit record for a mutation that never happened.
//
// The arm is a MEMBER OF A CONSTRUCTED TYPE, never a top-level function. `TriageArm()`
// can be stood up on its own and asked one question; `triageArm(…)` could only ever be
// reached through whatever called it. Applying the split rule honestly, `slice`,
// `result`, `now` and `sig` all vary per call, so NOTHING is constructor state and the
// constructor is empty — unlike the boundary's `ActionResolution(registry)`, which had
// a genuine invariant to hoist. An empty constructor is the correct answer here, not a
// missing one, and manufacturing a field to look busier would be the wrong lesson.

package adr.blocks.triage

import adr.contract.TriageEffect
import adr.contract.TriageResult
import adr.spine.pure.ArmOut
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp

internal class TriageArm {

    fun arm(
        slice: TriageSlice,
        result: TriageResult,
        now: Timestamp,
        sig: Signature,
    ): ArmOut<TriageSlice> = when (result) {
        is TriageResult.SetPriority -> {
            if (slice.tickets[result.ticket] == null) {
                ArmOut(
                    slice = slice,
                    notices = listOf(
                        Notice.Rejected(now, result.tool, "unknown ticket ${result.ticket.value}"),
                    ),
                )
            } else {
                ArmOut(
                    slice = slice.withPriority(result.ticket, result.level),
                    // `supersedes` is the fold reading ITS OWN current state (4.3): the tool
                    // returned raw inputs only, and could not have known the previous level.
                    effects = listOf(
                        TriageEffect.LogDecision(
                            at = now,
                            ticket = result.ticket,
                            level = result.level,
                            supersedes = slice.priority[result.ticket],
                            reason = result.reason,
                        ),
                    ),
                )
            }
        }
    }
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/triage/fold.ts
// ── blocks/triage/fold — the block's ARM (12.4) ────────────────────────────
// Three rules, mechanical, no exceptions:
//   1. the arm READS CURRENT STATE before it decides;
//   2. every effect push lives INSIDE the success branch;
//   3. a rejection folds a per-item Notice — never the session-global RunStatus,
//      never a mutation.
//
// The shipped reference failed all three: setPriority on an unknown ticket
// performed an Effect.Log, committed a SetPriority command, left folded state
// UNCHANGED — a clean-looking audit record for a mutation that never happened —
// and flipped the session banner to "degraded" for the rest of the session.
//
// The match is a `never`-guarded switch over the block's own sealed sub-union,
// so a new verb here fails to compile until its arm exists.

import type { Signature } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";
import { rejected } from "@adr/spine/pure/notice";
import type { ArmOut } from "@adr/spine/pure/verb";
import { armOut } from "@adr/spine/pure/verb";
import type { LogDecision, TriageResult } from "./contract";
import type { TriageSlice } from "./slice";
import { priorityOf, ticketOf, withPriority } from "./slice";

export function triageArm(
  slice: TriageSlice,
  r: TriageResult,
  now: Timestamp,
  _sig: Signature,
): ArmOut<TriageSlice> {
  switch (r.tool) {
    case "setPriority": {
      if (ticketOf(slice, r.ticket) === null) {
        return armOut(slice, [], [rejected(now, r.tool, `unknown ticket ${r.ticket}`)]);
      }
      // the fold — not the tool — derives `supersedes` from its own state
      const effect: LogDecision = {
        kind: "LogDecision",
        at: now,
        effectClass: "Routine",
        ticket: r.ticket,
        level: r.level,
        supersedes: priorityOf(slice, r.ticket),
        reason: r.reason,
      };
      return armOut(withPriority(slice, r.ticket, r.level), [effect], []);
    }
    default: {
      // The guard is on the DISCRIMINANT, not the value, because this block has
      // one verb today and a one-member "union" is just an interface — which TS
      // cannot narrow to `never`. Guarding `r.tool` is total either way: add
      // `archiveTicket` to the block's contract and this line stops compiling.
      const _never: never = r.tool;
      return _never;
    }
  }
}
